lib/spack/spack/cmd/providers.py

```diff
@@ -35,11 +36,46 @@
 
 
 def setup_parser(subparser):
+    subparser.epilog = 'If called without argument returns ' \
+                       'the list of all valid virtual packages'
     subparser.add_argument(
-        'vpkg_spec', metavar='VPACKAGE_SPEC', nargs=argparse.REMAINDER,
-        help='find packages that provide this virtual package')
+        'virtual_package',
+        nargs='*',
+        help='find packages that provide this virtual package'
+    )
 
 
 def providers(parser, args):
-    for spec in spack.cmd.parse_specs(args.vpkg_spec):
-        colify(sorted(spack.repo.providers_for(spec)), indent=4)
+    valid_virtuals = sorted(spack.repo.provider_index.providers.keys())
+
+    buffer = six.StringIO()
+    isatty = sys.stdout.isatty()
+    if isatty:
+        buffer.write('Virtual packages:\n')
+    colify.colify(valid_virtuals, output=buffer, tty=isatty, indent=4)
+    valid_virtuals_str = buffer.getvalue()
+
+    # If called without arguments, list all the virtual packages
+    if not args.virtual_package:
+        print(valid_virtuals_str)
+        return
+
+    # Otherwise, parse the specs from command line
+    specs = spack.cmd.parse_specs(args.virtual_package)
+
+    # Check prerequisites
+    non_virtual = [
+        str(s) for s in specs if not s.virtual or s.name not in valid_virtuals
+    ]
+    if non_virtual:
+        msg = 'non-virtual specs cannot be part of the query '
+        msg += '[{0}]\n'.format(', '.join(non_virtual))
+        msg += valid_virtuals_str
+        raise ValueError(msg)
+
+    # Display providers
+    for spec in specs:
+        if sys.stdout.isatty():
+            print("{0}:".format(spec))
+        spack.cmd.display_specs(sorted(spack.repo.providers_for(spec)))
```

Comment:
This is the only modification I didn't do. I think your solution will be ok as long as the list remains on one line, otherwise it will break. Don't know if it is the case to modify `display_specs` in this PR, leave things as they are or display the specs using another function.